### DISCUSSION TITLE
Exclude link targets when using strict mode?

### DISCUSSION THREAD

**Discussion by martinbira** [mkdocs/mkdocs#2957] [Ideas]
We've been running our MkDocs installation in `strict` mode since 2017 and it's such a great and easy way to check if your links are in order. However, we've now started to automatically generate parts of our documentation directly from our API and then injecting it in our build process before the rest of the site is generated.

This has resulted in the issue that all our links to those pages now fail when using `mkdocs serve` locally and we receive 70+ errors. As I see it, at the moment I have one option to retain local  `serve` functionality to keep using `strict` for build and internal link validation - convert all our links to absolute URLs as they are not affected by `strict`.

This will however cause issues in our docs development environment as we automatically create sites for internal review based on the active branch we use, and the links would point to our live documentation.

I've checked https://github.com/mkdocs/mkdocs/issues/1443 and adding a setting such as `exclude_targets` or similar will most likely not be added, and I haven't seem to find a plugin that have this functionality.

Does anyone have any suggestions for either a plugin (that I've missed) or any other method or workaround to still run `strict` mode to check my links on serve/build while not failing for a sub set of target pages?

**Comment by oprypin** [comment]
> generate parts of our documentation directly from our API and then injecting it in our build process

Side note: I am interested in which solution you arrived at, currently looking for ideas in this space and how to improve [mkdocstrings](https://github.com/mkdocstrings/mkdocstrings).

And does this mean that your API pages have a totally different style, not themed to mkdocs? Or how do you inject them? It's important.

**Comment by oprypin** [comment]
My advice is not a direct answer to the title, but instead a better suggestion:
Make MkDocs own the build end-to-end. So that there are no wrappers around MkDocs, no "pre-" steps, `mkdocs build` will just encompass everything.

I assume that you have some files originating from outside, that wasn't precisely specified.
As an optional suggestion, you can run the code that produces these files in some early [event](https://www.mkdocs.org/dev-guide/plugins/#events) - actually `on_files` fits fine.

**Then the core of the suggestion:**

The way that MkDocs [checks](https://github.com/mkdocs/mkdocs/blame/1.3.1/mkdocs/structure/pages.py#L220) for links' validity is checking if they're in the `Files` collection. So, you can [add](https://github.com/mkdocs/mkdocs/blob/1.3.1/mkdocs/structure/files.py#L34) a [new file](https://github.com/oprypin/mkdocs-gen-files/blob/faf8012228b3760f920e566b12aef696210d7ec2/mkdocs_gen_files/editor.py#L40) to that collection properly, instead of bypassing it.
This should be done in the `on_files` event.

```python
import os.path
import tempfile
from mkdocs.structure.files import File

my_src_dir = tempfile.TemporaryDirectory(prefix='my_src_dir_')

def on_files(files, config):
    generated = ['foo.html']

    for fn in generated:
        with open(os.path.join(my_src_dir.name, fn), 'w') as f:
            f.write("hi")

        f = File(
            fn,
            src_dir=my_src_dir.name,
            dest_dir=config['site_dir'],
            use_directory_urls=config['use_directory_urls'],
        )
        files.append(f)

    return files
```

To achieve that, you can do one of these:

* Write your own plugin
* Same as writing your own plugin but with less overhead:
  use https://github.com/aklajnert/mkdocs-simple-hooks (more flexible but you write more code)
* https://github.com/oprypin/mkdocs-gen-files
  (you avoid the intricacies of actually writing an `on_files` method, but you'll still need to manually `walk` the files in the source dir and copy them)


Since you're not providing an exact example of what you currently have, I am not writing a complete code example either, I don't want to guess and make up an example from scratch myself. Otherwise, would love to.

**Comment by martinbira** [comment]
Thanks for such a thorough answer even with our limited info!

So our solution was to create empty markdown-files in the repository for all our API endpoint pages that we would generate and need links for. A few index-pages with actual content was also created.

The generated markdown files (based on an OpenAPI specification) will then overwrite the empty endpoint-pages in our Jenkins build pipeline. At the same time we generate a new `mkdocs.yaml`-file to make sure that the newly generated are included in the navigation even though they don't have any links to them.

So a summary: 
Fetch Open API specification -> Generate pages for endpoints based on specification -> Insert into an MkDocs directort -> Continue with `mkdocs build` and the rest of the Pipeline.

Here's the result: https://docs.apptus.com/esales-4/integration/api/specification/

**Reply by oprypin** [reply]
OK so then my answer applies exactly, and I strongly recommend it. You don't have to physically generate a mkdocs.yml file (see below) or the stub file.

You can also directly get inspiration from this recipe, which generates files `foo.md` that just contain `::: Foo` but instead in your case you'd directly run your generator (either previously or in the same build step, which I recommend) and then copy your files by reading + writing with gen-files.
https://mkdocstrings.github.io/recipes/
Particularly, here's generating a sub-nav:
https://mkdocstrings.github.io/recipes/#generate-a-literate-navigation-file
